zkevm-circuits/src/state_circuit/binary_number.rs

```diff
@@ -0,0 +1,172 @@
+use crate::{
+    evm_circuit::{
+        table::RwTableTag,
+        util::{and, not},
+    },
+    util::Expr,
+};
+use eth_types::Field;
+use halo2_proofs::{
+    circuit::Region,
+    plonk::{Advice, Column, ConstraintSystem, Error, Expression, Fixed, VirtualCells},
+    poly::Rotation,
+};
+use std::collections::BTreeSet;
+use std::marker::PhantomData;
+use strum::IntoEnumIterator;
+
+pub trait AsBits<const N: usize> {
+    // Return the bits of self, starting from the most significant.
+    fn as_bits(&self) -> [bool; N];
+}
+
+impl AsBits<4> for RwTableTag {
+    fn as_bits(&self) -> [bool; 4] {
+        let mut bits = [false; 4];
+        let mut x = *self as u8;
+        for i in 0..4 {
+            bits[3 - i] = x % 2 == 1;
+            x /= 2;
+        }
+        bits
+    }
+}
+
+impl<const N: usize> AsBits<N> for usize {
+    fn as_bits(&self) -> [bool; N] {
+        let mut bits = [false; N];
+        let mut x = *self as u64;
+        for i in 0..N {
+            bits[N - 1 - i] = x % 2 == 1;
+            x /= 2;
+        }
+        bits
+    }
+}
+
+#[derive(Clone, Copy)]
+pub struct Config<T, const N: usize>
+where
+    T: AsBits<N>,
+{
+    pub bits: [Column<Advice>; N],
```

Comment:
Could you add this comment?
- Must be constrained to be binary for correctness.